File: src/main/rooms/PacksRoom.java
package main.rooms;

import main.buttons.Group;
import main.buttons.LongButton;
import main.buttons.Pack;
import main.buttons.iconbuttons.Exit;
import main.buttons.iconbuttons.IconButton;
import processing.core.PConstants;
import processing.data.JSONArray;
import processing.data.JSONObject;

import java.io.File;
import java.util.ArrayList;

public class PacksRoom extends Room {

    private static final int GAP = 100;

    private Exit exit;
    private IconButton done;

    private ArrayList<Pack> unloadedPacks = new ArrayList<>();
    private ArrayList<Pack> loadedPacks = new ArrayList<>();
    private Pack movePack; //pack to be moved

    public PacksRoom() {
        this.exit = new Exit();
        this.done = new IconButton("resources/images/done_button.png") {
            @Override
            public void clicked() {
                main.packsRoom.save();
                main.switchRoom(main.loading);
            }
        };
    }

    public void setMovePack(Pack movePack) {
        this.movePack = movePack;
    }

    @Override
    public void setup() {
        this.unloadedPacks.clear();
        this.loadedPacks.clear();
        this.movePack = null;

        File[] directory = new File("resources/packs/").listFiles();
        if (directory != null) {
            for (File folder : directory) {
                File jsonFile = new File(folder.getAbsolutePath() + "/pack.json");
                File icon = new File(folder.getAbsolutePath() + "/icon.png");
                if (jsonFile.exists() && icon.exists()) {
                    JSONObject object = main.loadJSONObject(jsonFile.getAbsolutePath());
                    if (object.getString("name").equals("Alchemy")) {
                        System.err.println("Error: pack name cannot be name \"Alchemy\"");
                        continue;
                    }
                    this.unloadedPacks.add(new Pack(folder.getAbsolutePath(), object));
                }
            }
        }
        this.unloadedPacks.add(new Pack()); //add default pack

        main.packsRoom.unloadedPacks.sort(Pack::compareTo);

        //loaded packs are read from settings file
        JSONArray array = main.getSettings().getJSONArray("loaded packs");
        for (int i = 0; i < array.size(); i++) {
            String string = array.getString(i);
            Pack pack;
            if (string.equals("Alchemy")) {
                pack = getPack("Alchemy", this.unloadedPacks);
            } else {
                pack = getPack(string, this.unloadedPacks);
                if (pack == null) {
                    System.err.println("Error: can't load pack with name: " + string);
                    continue;
                }
            }
            this.unloadedPacks.remove(pack);
            this.loadedPacks.add(pack);
        }
    }

    public ArrayList<Pack> getLoadedPacks() {
        return this.loadedPacks;
    }

    @Override
    public void draw() {
        this.drawTitle("packs", "packs");

        int x = main.screenWidth / 2 - GAP / 2 - LongButton.WIDTH;
        int y = 120;

        main.textSize(24);
        main.textAlign(PConstants.CENTER, PConstants.CENTER);
        main.fill(255);
        main.text(main.getLanguageSelected().getLocalizedString("packs", "unloaded packs"), x + LongButton.WIDTH / 2, y);

        y += 40;

        for (Pack pack : this.unloadedPacks) {
            pack.draw(x, y);
            y += pack.getHeight();
        }

        x = main.screenWidth / 2 + GAP / 2;
        y = 120;

        main.textSize(24);
        main.textAlign(PConstants.CENTER, PConstants.CENTER);
        main.fill(255);
        main.text(main.getLanguageSelected().getLocalizedString("packs", "loaded packs"), x + LongButton.WIDTH / 2, y);

        y += 40;

        for (Pack pack : this.loadedPacks) {
            pack.draw(x, y);
            y += pack.getHeight();
        }

        this.done.setDisabled(this.loadedPacks.size() == 0);
        this.done.draw(main.screenWidth - Group.GAP - IconButton.SIZE - (IconButton.SIZE + Group.GAP), main.screenHeight - Group.GAP - IconButton.SIZE);
        this.exit.draw();
    }

    @Override
    public void mousePressed() {
        for (Pack pack : this.unloadedPacks) {
            pack.mousePressed();
        }

        for (Pack pack : this.loadedPacks) {
            pack.mousePressed();
        }

        if (this.movePack != null) {
            if (main.packsRoom.unloadedPacks.contains(this.movePack)) {
                main.packsRoom.unloadedPacks.remove(this.movePack);
                main.packsRoom.loadedPacks.add(0, this.movePack);
            } else {
                main.packsRoom.loadedPacks.remove(this.movePack);
                main.packsRoom.unloadedPacks.add(this.movePack);
            }
            main.packsRoom.unloadedPacks.sort(Pack::compareTo);
        }
        this.movePack = null;

        this.done.mousePressed();
        this.exit.mousePressed();
    }

    public void save() {
        JSONArray array = new JSONArray();
        for (Pack pack : this.loadedPacks) {
            array.append(pack.getName());
        }
        main.getSettings().put("loaded packs", array);
        main.saveSettings();
    }

    private static Pack getPack(String name, ArrayList<Pack> list) {
        for (Pack pack : list) {
            if (pack.getName().equals(name)) {
                return pack;
            }
        }
        return null;
    }


}

File: src/main/rooms/PacksRoom.kt
package main.rooms

import main.Language
import main.Main
import main.buttons.Group
import main.buttons.LongButton
import main.buttons.Pack
import main.buttons.iconbuttons.Exit
import main.buttons.iconbuttons.IconButton
import processing.core.PConstants
import processing.data.JSONArray
import processing.data.JSONObject
import processing.data.StringList
import java.io.File

object PacksRoom : Room() {
    private const val GAP = 100
    private val exit: Exit = Exit()
    private val done: IconButton
    private val unloadedPacks = mutableListOf<Pack>()
    val loadedPacks = mutableListOf<Pack>()
    private var movePack: Pack? = null //pack to be moved

    init {
        done = object : IconButton("resources/images/done_button.png") {
            override fun clicked() {
                save()
                Main.switchRoom(Loading)
            }
        }
    }

    fun setMovePack(movePack: Pack?) {
        this.movePack = movePack
    }

    override fun setup() {
        unloadedPacks.clear()
        loadedPacks.clear()
        movePack = null
        File("resources/packs/").listFiles()?.let {
            for (folder in it) {
                val jsonFile = File(folder.absolutePath + "/pack.json")
                val icon = File(folder.absolutePath + "/icon.png")
                if (jsonFile.exists() && icon.exists()) {
                    val `object`: JSONObject = Main.loadJSONObject(jsonFile.absolutePath)
                    if (`object`.getString("name") == "Alchemy") {
                        System.err.println("Error: pack name cannot be name \"Alchemy\"")
                        continue
                    }
                    unloadedPacks.add(Pack(folder.absolutePath, `object`))
                }
            }
        }
        unloadedPacks.add(Pack()) //add default pack
        unloadedPacks.sort()

        //loaded packs are read from settings file
        for (string in Main.settings.getJSONArray("loaded packs").stringArray) {
            var pack: Pack?
            if (string == "Alchemy") {
                pack = getPack("Alchemy")
            } else {
                pack = getPack(string)
                if (pack == null) {
                    System.err.println("Error: can't load pack with name: $string")
                    continue
                }
            }
            unloadedPacks.remove(pack)
            loadedPacks.add(pack!!)
        }
    }

    override fun draw() {
        drawTitle("packs", "packs")
        var x = Main.screenWidth / 2 - GAP / 2 - LongButton.WIDTH
        var y = 120
        Main.textSize(24f)
        Main.textAlign(PConstants.CENTER, PConstants.CENTER)
        Main.fill(255)
        Main.text(Language.languageSelected.getLocalizedString("packs", "unloaded packs"), x + LongButton.WIDTH / 2f, y.toFloat())
        y += 40
        for (pack in unloadedPacks) {
            pack.draw(x.toFloat(), y.toFloat())
            y += pack.height
        }
        x = Main.screenWidth / 2 + GAP / 2
        y = 120
        Main.textSize(24f)
        Main.textAlign(PConstants.CENTER, PConstants.CENTER)
        Main.fill(255)
        Main.text(Language.languageSelected.getLocalizedString("packs", "loaded packs"), x + LongButton.WIDTH / 2f, y.toFloat())
        y += 40
        for (pack in loadedPacks) {
            pack.draw(x.toFloat(), y.toFloat())
            y += pack.height
        }
        done.disabled = loadedPacks.isEmpty()
        done.draw((Main.screenWidth - Group.GAP - IconButton.SIZE - (IconButton.SIZE + Group.GAP)).toFloat(), (Main.screenHeight - Group.GAP - IconButton.SIZE).toFloat())
        exit.draw()
    }

    override fun mousePressed() {
        for (pack in unloadedPacks) {
            pack.mousePressed()
        }
        for (pack in loadedPacks) {
            pack.mousePressed()
        }
        movePack?.let {
            if (unloadedPacks.contains(it)) {
                unloadedPacks.remove(it)
                loadedPacks.add(0, it)
            } else {
                loadedPacks.remove(it)
                unloadedPacks.add(it)
            }
            unloadedPacks.sort()
        }
        movePack = null
        done.mousePressed()
        exit.mousePressed()
    }

    private fun save() {
        Main.settings.put("loaded packs", JSONArray(StringList( loadedPacks.map { it.name })))
        Main.saveSettings()
    }

    private fun getPack(name: String): Pack? {
        for (pack in unloadedPacks) {
            if (pack.name == name) {
                return pack
            }
        }
        return null
    }
}